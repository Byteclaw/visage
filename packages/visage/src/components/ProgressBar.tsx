import {
  ExtractVisageComponentProps,
  markAsVisageComponent,
  VisageComponent,
} from '@byteclaw/visage-core';
import React from 'react';
import { createComponent } from '../core';
import { visuallyHiddenStyles } from './shared';

const HiddenProgressBar = createComponent('progress', {
  styles: visuallyHiddenStyles,
});

const ProgressBarBase = createComponent('div', {
  displayName: 'ProgressBar',
  styles: {
    fontSize: 'inherit',
    lineHeight: 'inherit',
    overflow: 'hidden',
    position: 'relative',
    width: '100%',
    '&:before': {
      backgroundColor: 'currentColor',
      content: '"\\200b"',
      display: 'inline-block',
      opacity: 0.2,
      width: '100%',
    },
  },
});

const ProgressBarProgress = createComponent('div', {
  displayName: 'ProgressBarProgress',
  styles: {
    '&::before': {
      backgroundColor: 'currentColor',
      content: '"\\200b"',
      display: 'inline-block',
      lineHeight: '100%',
      width: '100%',
    },
    height: '100%',
    left: 0,
    position: 'absolute',
    top: 0,
    maxWidth: '100%',
    willChange: 'width',
    transition: 'width .5s cubic-bezier(.64,0,.35,1)',
  },
});

export const ProgressBar: VisageComponent<{
  baseProps?: ExtractVisageComponentProps<typeof ProgressBarBase>;
  progressProps?: ExtractVisageComponentProps<typeof ProgressBarProgress>;
  max?: number;
  value: number;
  valueText?: string;
}> = function ProgressBar({
  baseProps,
  max = 100,
  progressProps,
  value,
  valueText,
  ...restProps
}: any) {
  return (
    <ProgressBarBase {...baseProps} {...restProps}>
      <HiddenProgressBar aria-valuetext={valueText} value={value} max={max} />
      <ProgressBarProgress
        style={{ width: `${(100 / max) * value}%` }}
        {...progressProps}
      />
    </ProgressBarBase>
  );
};

markAsVisageComponent(ProgressBar);
