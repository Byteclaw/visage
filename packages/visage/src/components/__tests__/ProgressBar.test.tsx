import React from 'react';
import { ProgressBar } from '../ProgressBar';
import { createTestTheme, render } from './render';

describe('ProgressBar', () => {
  it('works correctly', () => {
    const { container, getByTestId, rerender } = render(
      <ProgressBar progressProps={{ 'data-testid': 'p' } as any} value={0} />,
    );

    expect(container.querySelector('progress')).toHaveAttribute('max', '100');
    expect(container.querySelector('progress')).toHaveAttribute('value', '0');

    expect(getByTestId('p')).toHaveStyle({ width: '0%' });

    rerender(
      <ProgressBar progressProps={{ 'data-testid': 'p' } as any} value={50} />,
    );

    expect(container.querySelector('progress')).toHaveAttribute('max', '100');
    expect(container.querySelector('progress')).toHaveAttribute('value', '50');

    expect(getByTestId('p')).toHaveStyle({ width: '50%' });
  });

  it('works with max prop', () => {
    const { container, getByTestId, rerender } = render(
      <ProgressBar
        progressProps={{ 'data-testid': 'p' } as any}
        max={200}
        value={0}
      />,
    );

    expect(container.querySelector('progress')).toHaveAttribute('max', '200');
    expect(container.querySelector('progress')).toHaveAttribute('value', '0');

    expect(getByTestId('p')).toHaveStyle({ width: '0%' });

    rerender(
      <ProgressBar
        progressProps={{ 'data-testid': 'p' } as any}
        max={200}
        value={50}
      />,
    );

    expect(container.querySelector('progress')).toHaveAttribute('max', '200');
    expect(container.querySelector('progress')).toHaveAttribute('value', '50');

    expect(getByTestId('p')).toHaveStyle({ width: '25%' });
  });

  it('renders correctly', () => {
    const { asFragment } = render(<ProgressBar value={0} />);

    expect(asFragment()).toMatchSnapshot();
  });

  it('is extendable by ProgressBar face', () => {
    const { asFragment } = render(<ProgressBar value={0} />, {
      ds: {
        theme: createTestTheme({
          ProgressBar: {
            color: 'pink',
          },
        }),
      },
    });

    expect(asFragment()).toMatchSnapshot();
  });

  it('is extendable by ProgressBarProgress face', () => {
    const { asFragment } = render(<ProgressBar value={0} />, {
      ds: {
        theme: createTestTheme({
          ProgressBarProgress: {
            color: 'pink',
            fontSize: -1,
            lineHeight: -1,
          },
        }),
      },
    });

    expect(asFragment()).toMatchSnapshot();
  });
});
