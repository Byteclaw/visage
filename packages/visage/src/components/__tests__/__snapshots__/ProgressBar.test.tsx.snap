// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ProgressBar is extendable by ProgressBar face 1`] = `
<DocumentFragment>
  .emotion-0 {
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  position: fixed;
  bottom: 0;
  left: 0;
  right: 0;
  z-index: 600;
}

<div
    aria-live="polite"
    class="emotion-0"
    data-toastmanager="true"
  />
  .emotion-0 {
  border: 0;
  -webkit-clip: rect(0,0,0,0);
  clip: rect(0,0,0,0);
  height: 1px;
  overflow: hidden;
  margin: -1px;
  padding: 0px;
  position: absolute;
  white-space: nowrap;
  width: 1px;
}

.emotion-1 {
  height: 100%;
  left: 0;
  position: absolute;
  top: 0;
  max-width: 100%;
  will-change: width;
  -webkit-transition: width .5s cubic-bezier(.64,0,.35,1);
  transition: width .5s cubic-bezier(.64,0,.35,1);
}

.emotion-1::before {
  background-color: currentColor;
  content: "\\200b";
  display: inline-block;
  line-height: 100%;
  width: 100%;
}

.emotion-2 {
  font-size: inherit;
  line-height: inherit;
  overflow: hidden;
  position: relative;
  width: 100%;
  color: rgb(255,192,203);
}

.emotion-2:before {
  background-color: currentColor;
  content: "\\200b";
  display: inline-block;
  opacity: 0.2;
  width: 100%;
}

<div
    class="emotion-2"
  >
    <progress
      class="emotion-0"
      max="100"
      value="0"
    />
    <div
      class="emotion-1"
      style="width: 0%;"
    />
  </div>
</DocumentFragment>
`;

exports[`ProgressBar is extendable by ProgressBarProgress face 1`] = `
<DocumentFragment>
  .emotion-0 {
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  position: fixed;
  bottom: 0;
  left: 0;
  right: 0;
  z-index: 600;
}

<div
    aria-live="polite"
    class="emotion-0"
    data-toastmanager="true"
  />
  .emotion-2 {
  font-size: inherit;
  line-height: inherit;
  overflow: hidden;
  position: relative;
  width: 100%;
}

.emotion-2:before {
  background-color: currentColor;
  content: "\\200b";
  display: inline-block;
  opacity: 0.2;
  width: 100%;
}

.emotion-0 {
  border: 0;
  -webkit-clip: rect(0,0,0,0);
  clip: rect(0,0,0,0);
  height: 1px;
  overflow: hidden;
  margin: -1px;
  padding: 0px;
  position: absolute;
  white-space: nowrap;
  width: 1px;
}

.emotion-1 {
  height: 100%;
  left: 0;
  position: absolute;
  top: 0;
  max-width: 100%;
  will-change: width;
  -webkit-transition: width .5s cubic-bezier(.64,0,.35,1);
  transition: width .5s cubic-bezier(.64,0,.35,1);
  color: rgb(255,192,203);
  font-size: 10px;
  line-height: 24px;
}

.emotion-1::before {
  background-color: currentColor;
  content: "\\200b";
  display: inline-block;
  line-height: 100%;
  width: 100%;
}

<div
    class="emotion-2"
  >
    <progress
      class="emotion-0"
      max="100"
      value="0"
    />
    <div
      class="emotion-1"
      style="width: 0%;"
    />
  </div>
</DocumentFragment>
`;

exports[`ProgressBar renders correctly 1`] = `
<DocumentFragment>
  .emotion-0 {
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  position: fixed;
  bottom: 0;
  left: 0;
  right: 0;
  z-index: 600;
}

<div
    aria-live="polite"
    class="emotion-0"
    data-toastmanager="true"
  />
  .emotion-2 {
  font-size: inherit;
  line-height: inherit;
  overflow: hidden;
  position: relative;
  width: 100%;
}

.emotion-2:before {
  background-color: currentColor;
  content: "\\200b";
  display: inline-block;
  opacity: 0.2;
  width: 100%;
}

.emotion-0 {
  border: 0;
  -webkit-clip: rect(0,0,0,0);
  clip: rect(0,0,0,0);
  height: 1px;
  overflow: hidden;
  margin: -1px;
  padding: 0px;
  position: absolute;
  white-space: nowrap;
  width: 1px;
}

.emotion-1 {
  height: 100%;
  left: 0;
  position: absolute;
  top: 0;
  max-width: 100%;
  will-change: width;
  -webkit-transition: width .5s cubic-bezier(.64,0,.35,1);
  transition: width .5s cubic-bezier(.64,0,.35,1);
}

.emotion-1::before {
  background-color: currentColor;
  content: "\\200b";
  display: inline-block;
  line-height: 100%;
  width: 100%;
}

<div
    class="emotion-2"
  >
    <progress
      class="emotion-0"
      max="100"
      value="0"
    />
    <div
      class="emotion-1"
      style="width: 0%;"
    />
  </div>
</DocumentFragment>
`;
